### PR TITLE
COSBETA-177: Invite user - email address field validation issues

### PR DESCRIPTION
**Pull request by aidayounesi**
## Description
Validate email address using a regex both in the controller and the model

## Checklist:

- [ ] Automated checks are passing locally.
- [ ] CHANGELOG updated if change is worth telling users about.
### General testing
- [ ] Test without javascript
- [ ] Test on small screen
### Accessibility testing
- [ ] Works keyboard only
- [ ] Tested with one screen reader
- [ ] Zoom page to 400% - content still visible
- [ ] Disable css - does content make sense and appear in a logical order?
- [ ] Passes automated checker (automated in build or manual)

### CODE REVIEW

File: cosmetics-web/app/controllers/responsible_persons/team_members_controller.rb [hunk 1]
```diff
@@ -8,6 +8,11 @@ def create
       return redirect_to new_responsible_person_team_member_path(@responsible_person, blank_input: true)
     end
 
+    # Check the user entered a valid email address
```
Comment (TiagoSusanoPinto) [inline cosmetics-web/app/controllers/responsible_persons/team_members_controller.rb]: Do we need this comment? Should we extract the validation to function, instead?
e.g.:
`def is_valid_email email_address`
`  email_address =~ ``URI::MailTo::EMAIL_REGEXP`` `
`end`

I've notice that all validations on this method have a comment, but should we extract all of those validations to separated methods?

Comment (TiagoSusanoPinto) [reply cosmetics-web/app/controllers/responsible_persons/team_members_controller.rb]: Also, should we use the activeRecord's validation instead? (not sure if possible)

File: cosmetics-web/app/controllers/responsible_persons/team_members_controller.rb [hunk 2]
```diff
@@ -8,6 +8,11 @@ def create
...
+    if not team_member_params[:email_address] =~ URI::MailTo::EMAIL_REGEXP
```
Comment (TiagoSusanoPinto) [inline cosmetics-web/app/controllers/responsible_persons/team_members_controller.rb]: I've notice that `team_member_params[:email_address]` is repeated several time in this method.
Should you create a variable in the begging of the function with it?
This change might even increase the readability, so you could just remove some comments instead :)

File: cosmetics-web/app/controllers/responsible_persons/team_members_controller.rb (hunk 1 above)
Comment (tom-bell-softwire) [reply cosmetics-web/app/controllers/responsible_persons/team_members_controller.rb]: Great point, Tiago. I think the plan is to change the form in the view to use a model instance, so we can make use of the standard validation and error message handling.

File: cosmetics-web/app/controllers/responsible_persons/team_members_controller.rb [hunk 3]
```diff
@@ -1,34 +1,30 @@
...
-      "email_address = ? AND responsible_person_id = ? AND expires_at > ?",
-      User.current.email,
-      params[:responsible_person_id],
-      DateTime.current
-)
-
-    if pending_responsible_person_user.any?
+    pending_requests = PendingResponsiblePersonUser.pending_requests_to_join_responsible_person(
+      User.current,
+        @responsible_person
```
Comment (tom-bell-softwire) [inline cosmetics-web/app/controllers/responsible_persons/team_members_controller.rb]: Indentation looks a bit off here.

File: cosmetics-web/app/controllers/responsible_persons/team_members_controller.rb [hunk 4]
```diff
@@ -42,9 +38,30 @@ def set_responsible_person
...
 
   def team_member_params
-    params.require(:team_member)
-      .permit(
-        :email_address
-      )
+    team_member_session_params.merge(team_member_request_params)
+  end
+
+  def team_member_session_params
```
[the remark this answers is not in the document]
Comment (aidayounesi) [reply cosmetics-web/app/controllers/responsible_persons/team_members_controller.rb]: You are right, no need for session. thanks

File: cosmetics-web/app/controllers/responsible_persons/team_members_controller.rb [hunk 5]
```diff
@@ -42,9 +38,30 @@ def set_responsible_person
...
+    session.fetch(:team_member, {})
+  end
+
+  def team_member_request_params
+    params.fetch(:team_member, {}).permit(
```
[the remark this answers is not in the document]
Comment (aidayounesi) [reply cosmetics-web/app/controllers/responsible_persons/team_members_controller.rb]: Sure, thanks

File: cosmetics-web/app/controllers/responsible_persons/team_members_controller.rb [hunk 6]
```diff
@@ -1,34 +1,30 @@
...
-      return redirect_to new_responsible_person_team_member_path(@responsible_person, blank_input: true)
-    end
+  def new; end
 
-    # Check if the user is already a member of the team
-    if @responsible_person.responsible_person_users.any? { |user| user.email_address == team_member_params[:email_address] }
-      return redirect_to new_responsible_person_team_member_path(@responsible_person, user_already_exists: true)
+  def create
+    if responsible_person_saved?
+      NotifyMailer.send_responsible_person_invite_email(@responsible_person.id, @responsible_person.name,
```
[the remark this answers is not in the document]
Comment (aidayounesi) [reply cosmetics-web/app/controllers/responsible_persons/team_members_controller.rb]: Sure, thanks

File: cosmetics-web/app/controllers/responsible_persons/team_members_controller.rb [hunk 7]
```diff
@@ -42,9 +38,30 @@ def set_responsible_person
...
+      :email_address
+    )
+  end
+
+  def set_team_member
+    @team_member = @responsible_person.pending_responsible_person_users.build(team_member_params)
+  end
+
+  def responsible_person_saved?
+    return false unless responsible_person_valid?
```
[the remark this answers is not in the document]
Comment (aidayounesi) [reply cosmetics-web/app/controllers/responsible_persons/team_members_controller.rb]: Sure, thanks

File: cosmetics-web/app/views/responsible_persons/team_members/new.html.slim [hunk 8]
```diff
@@ -1,23 +1,21 @@
...
-    - if params[:user_already_exists]
-      span.govuk-error-message
-        | The user is already a member of this team
-    - if params[:blank_input]
-      span.govuk-error-message
-        | Enter an email address
-    p Inviting a team member will allow them to:
-    ul.govuk-list.govuk-list--bullet
+
+    = form_with model: @team_member, scope: :team_member,
```
[the remark this answers is not in the document]
Comment (aidayounesi) [reply cosmetics-web/app/views/responsible_persons/team_members/new.html.slim]: > Is `scope` is required here if you specify a `model` ?
It doesn't work if I remove the scope

**Pull request merged**